src/appkit.cs

```diff
@@ -4705,27 +4832,82 @@ partial interface NSControl {
 		void EndEditing ([NullAllowed] NSText textObj);
 	}
 
-	[DesignatedDefaultCtor]
-	[BaseType (typeof (NSObject))]
-	interface NSController : NSCoding {
+	[Protocol]
+	interface NSEditorRegistration
+	{
 		[Export ("objectDidBeginEditing:")]
-		void ObjectDidBeginEditing (NSObject editor);
+		void ObjectDidBeginEditing (INSEditor editor);
 
 		[Export ("objectDidEndEditing:")]
-		void ObjectDidEndEditing (NSObject editor);
+		void ObjectDidEndEditing (INSEditor editor);
+	}
+
+	[Category]
+	[BaseType (typeof(NSObject))]
+	interface NSObject_NSEditorRegistration
+	{
+		[Export ("objectDidBeginEditing:")]
+		void ObjectDidBeginEditing (INSEditor editor);
+
+		[Export ("objectDidEndEditing:")]
+		void ObjectDidEndEditing (INSEditor editor);
+	}
+	
+	interface INSEditor {}
 
+	[Protocol]
+	interface NSEditor
+	{
+		[Abstract]
 		[Export ("discardEditing")]
 		void DiscardEditing ();
 
+		[Abstract]
 		[Export ("commitEditing")]
-		bool CommitEditing { get; }
+		bool CommitEditing (); 
+
+		[Abstract]
+		[Export ("commitEditingWithDelegate:didCommitSelector:contextInfo:")]
+		void CommitEditing ([NullAllowed] NSObject delegateObject, [NullAllowed] Selector didCommitSelector, IntPtr contextInfo);
+
+		[Mac (10,7)]
+		[Abstract]
+		[Export ("commitEditingAndReturnError:")]
+		bool CommitEditing ([NullAllowed] out NSError error);
+	}
+
+	[DesignatedDefaultCtor]
+	[BaseType (typeof (NSObject))]
+	interface NSController : NSCoding, NSEditorRegistration 
+#if XAMCORE_4_0
+	, NSEditor // Conflict over if CommitEditing is a property or a method. NSViewController has it right so can't "fix" NSEditor to match existing API
+#endif
+	{
+		[Export ("discardEditing")]
+		void DiscardEditing ();
 
 		[Export ("commitEditingWithDelegate:didCommitSelector:contextInfo:")]
-		void CommitEditingWithDelegate (NSObject delegate1, Selector didCommitSelector, IntPtr contextInfo);
+		void CommitEditingWithDelegate ([NullAllowed] NSObject delegate1, [NullAllowed] Selector didCommitSelector, IntPtr contextInfo);
+
+#if XAMCORE_4_0
+		[Export ("objectDidBeginEditing:")]
+		void ObjectDidBeginEditing (INSEditor editor);
+
+		[Export ("objectDidEndEditing:")]
+		void ObjectDidEndEditing (INSEditor editor);
+#else
+		[Export ("objectDidBeginEditing:")]
+		void ObjectDidBeginEditing (NSObject editor);
+
+		[Export ("objectDidEndEditing:")]
+		void ObjectDidEndEditing (NSObject editor);
+#endif
+
+		[Export ("commitEditing")]
+		bool CommitEditing { get; }
```

Comment:
should not the `CommitEditing ();` be using for `XAMCORE_4_0` ?

Reply:
See above comment.